grpc-sys/build.rs

```diff
@@ -12,14 +12,190 @@
+                     .and_then(|status| if status.success() {
+                                   Ok(())
+                               } else {
+                                   Err(format!("wget exit with {}", status))
+                               }));
+
+            try!(Command::new("tar")
+                     .args(&["zxf", &tgz_file_name])
+                     .current_dir(&out_dir)
+                     .status()
+                     .map_err(|err| format!("failed to execute tar: {}", err))
+                     .and_then(|status| if status.success() {
+                                   Ok(())
+                               } else {
+                                   Err(format!("tar exit with {}", status))
+                               }));
+            // clean base dir
+            try!(Command::new("rm")
+                     .args(&["-r", &self.to_dir])
+                     .current_dir(&out_dir)
+                     .status()
+                     .map_err(|err| format!("failed to execute tar: {}", err)));
+
+            Command::new("mv")
+                .args(&[&self.base_name, &self.to_dir])
+                .current_dir(&out_dir)
+                .status()
+                .map_err(|err| format!("mv execute failed: {}", err))
+                .and_then(|status| if status.success() {
+                              Ok(())
+                          } else {
+                              Err(format!("mv exit with {}", status))
+                          })
+        }
+    }
+
+
+    pub fn build_or_link_grpc(cc: &mut ::gcc::Config) {
+        let out_dir = env::var("OUT_DIR").expect("Can't access OUT_DIR");
+        DirBuilder::new()
+            .recursive(true)
+            .create(format!("{}/{}", out_dir, "grpc"))
+            .unwrap();
+
+        FakeGitSubmodule::new_with_semver("grpc", "grpc", GRPC_VERSION, "grpc")
+            .init()
+            .unwrap();
+
+        FakeGitSubmodule::new_with_semver("madler", "zlib", ZLIB_VERSION, "grpc/third_party/zlib")
+            .init()
+            .unwrap();
+
+        FakeGitSubmodule::new_with_hash("google",
+                                        "boringssl",
+                                        BORINGSSL_COMMIT_HASH,
+                                        "grpc/third_party/boringssl")
+                .init()
+                .unwrap();
+
+        FakeGitSubmodule::new_with_tag("c-ares",
+                                       "c-ares",
+                                       CARES_TAG,
+                                       "grpc/third_party/cares/cares")
+                .init()
+                .unwrap();
+
+        // fix multiple _main symbols
+        let _ = Command::new("rm")
+            .current_dir(&format!("{}/grpc/third_party/cares/cares", out_dir))
+            .args(&["acountry.c", "adig.c", "ahost.c"])
+            .status();
+
+        let dst = ::cmake::Config::new(format!("{}/grpc", out_dir))
```

Comment:
This is the style guidelines we follow: prefer fully importing types/traits while module-qualifying functions.